engine/table/src/main/java/io/deephaven/engine/table/impl/sources/regioned/RegionedColumnSourceManager.java

```diff
@@ -767,28 +769,44 @@ public long estimatePushdownFilterCost(
             final RowSet selection,
             final RowSet fullSet,
             final boolean usePrev,
-            final FilterContext context) {
+            final PushdownFilterContext context) {
 
         if (includedLocations().isEmpty()) {
             return 0; // No locations, no cost
         }
 
         // Test a few locations and assume the rest are similar (or no worse)
-        final long costPerLocation = includedLocations().stream().limit(PUSH_DOWN_LOCATIONS_TO_TEST).mapToLong(
-                location -> location.estimatePushdownFilterCost(filter, selection, fullSet, usePrev, context))
-                .reduce(Long::max).orElse(Long.MAX_VALUE);
+        final Collection<TableLocation> locations = includedLocations();
+        final long locationCost;
+
+        if (locations instanceof ArrayList) {
+            // We have fast access to the locations, so test a few regularly spaced locations and return the minimum
+            // cost encountered.
+            final ArrayList<TableLocation> locationList = (ArrayList<TableLocation>) locations;
+            final int step = Math.max(1, locationList.size() / PUSHDOWN_LOCATION_SAMPLES);
+            final Collection<TableLocation> candidateLocations =
+                    IntStream.range(0, Math.min(locationList.size(), PUSHDOWN_LOCATION_SAMPLES))
```

Comment:
You've got a stream no sense in turning it into an array list instead of just extending your pipeline with a .parallel() followed by the mapToLong

Reply:
we could also declare the stream outside the if, then have the last two lines with mapToLong and min shared.